implement plugin authentication, finer-grained restrictions/validation.

klihub opened the pull request:
This PR implements plugin authentication. The sole purpose of authenticated plugins is to allow finer-grained restrictions of selected NRI features than just a global on/off. This PR also updates the builtin default validator to allow finer-grained restrictions using overrides for authenticated plugins. 

In particular the patch series
- defines the wire protocol bits for authentication
- implements a challenge-response authentication mostly based on the ideas in #152
- implements runtime and plugin side support for authentication
- implements overrides for authenticated plugins in the default validator

Here are working trees of updated runtimes with support for authenticated NRI plugins and finer-grained validation using an NRI tree with this PR:
- containerd/main: https://github.com/klihub/containerd/tree/devel/nri-plugin-authentication
- cri-o/main: https://github.com/klihub/cri-o/tree/devel/nri-plugin-authentication

klihub committed:
auth: add authentication service and implementation.

Add a service and protocol for challenge-response authentication.
Define an interface for authentication implementations. Implement
a default authentication, mostly based on the ideas described in
https://github.com/containerd/nri/issues/152, the proposal for
plugin authentication. Add examples/keygen.go for generating key
pairs for this default implementation.

Signed-off-by: Krisztian Litkey <krisztian.litkey@intel.com>

klihub committed:
stub: implement plugin authentication.

Add a new option for acquiring a public/private key pair from
an external source for authenticating with the runtime. Using
this option implicitly enables plugin authentication prior to
registration.

Also add a new environment variable to read keys from files.
Together with bind-mounted secrets, this should provide enough
plumbing to enable transparent authentication of containerized
plugins.

Signed-off-by: Krisztian Litkey <krisztian.litkey@intel.com>

klihub committed:
Makefile: don't assume/hardcode proto location.

Neither assume that we have a single proto definition
nor hardcode proto file locations for post-processing
actions based on such an assumption.

Signed-off-by: Krisztian Litkey <krisztian.litkey@intel.com>

klihub committed:
api: pass authentication information to validation.

Add optional authentication-related information to the list
of plugin instances for validation. This should allow finer-
grained validation including loser or stricter restrictions
for authenticated vs. unauthenticated plugins.

Signed-off-by: Krisztian Litkey <krisztian.litkey@intel.com>

klihub committed:
default-validator: implement finer-grained validation.

Implement role based configuration overrides for authenticated
plugins. This allows setting up restrictive defaults but allow
a more permissive configuration for some or all authenticated
plugins based on their roles established during authentication.

Signed-off-by: Krisztian Litkey <krisztian.litkey@intel.com>

klihub committed:
contrib: add sample overlay for plugin authentication.

Signed-off-by: Krisztian Litkey <krisztian.litkey@intel.com>

klihub committed:
docs: update README.md.

Update documentation for plugin authentication and validation.

Signed-off-by: Krisztian Litkey <krisztian.litkey@intel.com>

chrishenzie commented on pkg/adaptation/plugin.go:
```diff
@@ -428,6 +447,74 @@ func (p *plugin) qualifiedName() string {
+	}
+
+	p.role.algo = algo
+	chal, key, err := p.role.algo.Challenge([]byte("XXtmpXXX"), req.PublicKey)
```
Judging by the name, I'm assuming this will be replaced? Should this be a random seed?

klihub replied:
Argh... that was already supposed to be replaced by a random seed. Thanks for spotting it !

klihub committed:
adaptation: implement plugin authentication.

Add support for plugin authentication. Authenticating a plugin
associates it with a role. A role has a unique name and it can
have an optional set of associated tags. These tags are opaque,
carrying no semantic meaning for authentication or NRI itself.

However, tags can have associated semantics during validation
and authorize plugins to perform otherwise restricted actions.
This is the primary intended usage for tags: allow validators
to associate rights with tags instead of role names.

Signed-off-by: Krisztian Litkey <krisztian.litkey@intel.com>
